realm/realm-library/src/androidTest/java/io/realm/ObjectChangeSetTests.java

```diff
@@ -0,0 +1,316 @@
+
+    @Rule
+    public final RunInLooperThread looperThread = new RunInLooperThread();
+
+    @Before
+    public void setUp() {
+    }
+
+    @After
+    public void tearDown() {
+    }
+
+    public static class PopulateOneAllTypes implements RunInLooperThread.RunnableBefore {
+
+        @Override
+        public void run(RealmConfiguration realmConfig) {
+            Realm realm = Realm.getInstance(realmConfig);
+            realm.executeTransaction(new Realm.Transaction() {
+                @Override
+                public void execute(Realm realm) {
+                    AllTypes allTypes = realm.createObject(AllTypes.class);
+                    allTypes.setColumnRealmObject(realm.createObject(Dog.class));
+                    allTypes.getColumnRealmList().add(realm.createObject(Dog.class));
+                }
+            });
+            realm.close();
+        }
+    }
+
+    private void checkDeleted(AllTypes allTypes) {
+        allTypes.addChangeListener(new RealmObjectChangeListener<AllTypes>() {
+            @Override
+            public void onChange(AllTypes object, ObjectChangeSet changeSet) {
+                assertEquals(0, changeSet.getChangedFields().length);
+                assertFalse(object.isValid());
+                assertTrue(changeSet.isDeleted());
+                looperThread.testComplete();
+            }
+        });
+        looperThread.keepStrongReference.add(allTypes);
+    }
+
+    private void checkChangedField(AllTypes allTypes, final String... fieldNames) {
+        assertNotNull(fieldNames);
+        allTypes.addChangeListener(new RealmObjectChangeListener<RealmModel>() {
+            @Override
+            public void onChange(RealmModel object, ObjectChangeSet changeSet) {
+                assertEquals(fieldNames.length, changeSet.getChangedFields().length);
+                for (String name : fieldNames) {
+                    assertTrue(changeSet.isFieldChanged(name));
+                    assertFalse(changeSet.isFieldChanged(name + "NotThere"));
+                    for (String field : changeSet.getChangedFields()) {
+                        if (field.equals(name)) {
+                            break;
+                        }
+                        fail("Cannot find field " + name + " in field changes.");
+                    }
+                }
+                looperThread.testComplete();
+            }
+        });
+        looperThread.keepStrongReference.add(allTypes);
+    }
+
+    private void listenerShouldNotBeCalled(AllTypes allTypes) {
+        allTypes.addChangeListener(new RealmObjectChangeListener<RealmModel>() {
+            @Override
+            public void onChange(RealmModel object, ObjectChangeSet changeSet) {
+                fail();
+            }
+        });
+        looperThread.postRunnableDelayed(new Runnable() {
+            @Override
+            public void run() {
+                looperThread.testComplete();
+            }
+        }, 100);
```

Comment:
Technically calling listeners for local commit in `commitTransaction()` will fix some silly bugs like https://github.com/realm/realm-java/issues/4384 so it is a good thing.